### ISSUE TITLE
Runtime error: TypeError: context must be specified

### ISSUE THREAD

**Issue by tritemio** [lmfit/lmfit-py#221]
In current master (b8e8a10e83156ff45493975c97a37af0a7f7699e), I get this error from an ipython notebook that works with 0.8.3.

I just dump the trace for now, I'll came back to it later:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-33-d9213941ff62> in <module>()
----> 1 ds.E_fitter.fit_histogram(mfit.factory_three_gaussians())

C:\Users\laser2002j\Anaconda\lib\site-packages\fretbursts-0.4rc9_9_g958ffd2-py2.7-win-amd64.egg\fretbursts\mfit.pyc in fit_histogram(self, model, pdf, **fit_kwargs)
    427         for i, data in enumerate(data_list):
    428             self.fit_res.append(
--> 429                 self.model.fit(data, x=self.hist_axis, **fit_kwargs)
    430                 )
    431             self.params.iloc[i] = pd.Series(self.fit_res[-1].values)

C:\Users\laser2002j\Anaconda\lib\site-packages\lmfit-0.8.3_54_gb8e8a10-py2.7.egg\lmfit\model.pyc in fit(self, data, params, weights, method, iter_cb, scale_covar, verbose, fit_kws, **kwargs)
    556 
    557         output = ModelFit(self, params, method=method, iter_cb=iter_cb,
--> 558                           scale_covar=scale_covar, fcn_kws=kwargs, **fit_kws)
    559         output.fit(data=data, weights=weights)
    560         output.components = self.components

C:\Users\laser2002j\Anaconda\lib\site-packages\lmfit-0.8.3_54_gb8e8a10-py2.7.egg\lmfit\model.pyc in __init__(self, model, params, data, weights, method, fcn_args, fcn_kws, iter_cb, scale_covar, **fit_kws)
    741         self.weights = weights
    742         self.method = method
--> 743         self.init_params = deepcopy(params)
    744         Minimizer.__init__(self, model._residual, params, fcn_args=fcn_args,
    745                            fcn_kws=fcn_kws, iter_cb=iter_cb,

C:\Users\laser2002j\Anaconda\lib\copy.pyc in deepcopy(x, memo, _nil)
    172             copier = getattr(x, "__deepcopy__", None)
    173             if copier:
--> 174                 y = copier(memo)
    175             else:
    176                 reductor = dispatch_table.get(cls)

C:\Users\laser2002j\Anaconda\lib\site-packages\lmfit-0.8.3_54_gb8e8a10-py2.7.egg\lmfit\parameter.pyc in __deepcopy__(self, memo)
     46         _pars = Parameters()
     47         for key, val in self._asteval.symtable.items():
---> 48             _pars._asteval.symtable[key] = deepcopy(val)
     49         for key, par in self.items():
     50             if isinstance(par, Parameter):

C:\Users\laser2002j\Anaconda\lib\copy.pyc in deepcopy(x, memo, _nil)
    161     copier = _deepcopy_dispatch.get(cls)
    162     if copier:
--> 163         y = copier(x, memo)
    164     else:
    165         try:

C:\Users\laser2002j\Anaconda\lib\copy.pyc in _deepcopy_method(x, memo)
    262 
    263 def _deepcopy_method(x, memo): # Copy instance methods
--> 264     return type(x)(x.im_func, deepcopy(x.im_self, memo), x.im_class)
    265 _deepcopy_dispatch[types.MethodType] = _deepcopy_method
    266 

C:\Users\laser2002j\Anaconda\lib\copy.pyc in deepcopy(x, memo, _nil)
    161     copier = _deepcopy_dispatch.get(cls)
    162     if copier:
--> 163         y = copier(x, memo)
    164     else:
    165         try:

C:\Users\laser2002j\Anaconda\lib\copy.pyc in _deepcopy_inst(x, memo)
    296     else:
    297         state = x.__dict__
--> 298     state = deepcopy(state, memo)
    299     if hasattr(y, '__setstate__'):
    300         y.__setstate__(state)

C:\Users\laser2002j\Anaconda\lib\copy.pyc in deepcopy(x, memo, _nil)
    161     copier = _deepcopy_dispatch.get(cls)
    162     if copier:
--> 163         y = copier(x, memo)
    164     else:
    165         try:

C:\Users\laser2002j\Anaconda\lib\copy.pyc in _deepcopy_dict(x, memo)
    255     memo[id(x)] = y
    256     for key, value in x.iteritems():
--> 257         y[deepcopy(key, memo)] = deepcopy(value, memo)
    258     return y
    259 d[dict] = _deepcopy_dict

C:\Users\laser2002j\Anaconda\lib\copy.pyc in deepcopy(x, memo, _nil)
    188                             raise Error(
    189                                 "un(deep)copyable object of type %s" % cls)
--> 190                 y = _reconstruct(x, rv, 1, memo)
    191 
    192     memo[d] = y

C:\Users\laser2002j\Anaconda\lib\copy.pyc in _reconstruct(x, info, deep, memo)
    332     if state:
    333         if deep:
--> 334             state = deepcopy(state, memo)
    335         if hasattr(y, '__setstate__'):
    336             y.__setstate__(state)

C:\Users\laser2002j\Anaconda\lib\copy.pyc in deepcopy(x, memo, _nil)
    161     copier = _deepcopy_dispatch.get(cls)
    162     if copier:
--> 163         y = copier(x, memo)
    164     else:
    165         try:

C:\Users\laser2002j\Anaconda\lib\copy.pyc in _deepcopy_dict(x, memo)
    255     memo[id(x)] = y
    256     for key, value in x.iteritems():
--> 257         y[deepcopy(key, memo)] = deepcopy(value, memo)
    258     return y
    259 d[dict] = _deepcopy_dict

C:\Users\laser2002j\Anaconda\lib\copy.pyc in deepcopy(x, memo, _nil)
    188                             raise Error(
    189                                 "un(deep)copyable object of type %s" % cls)
--> 190                 y = _reconstruct(x, rv, 1, memo)
    191 
    192     memo[d] = y

C:\Users\laser2002j\Anaconda\lib\copy.pyc in _reconstruct(x, info, deep, memo)
    327     if deep:
    328         args = deepcopy(args, memo)
--> 329     y = callable(*args)
    330     memo[id(x)] = y
    331 

C:\Users\laser2002j\Anaconda\lib\copy_reg.pyc in __newobj__(cls, *args)
     91 
     92 def __newobj__(cls, *args):
---> 93     return cls.__new__(cls, *args)
     94 
     95 def _slotnames(cls):

zmq/backend/cython/socket.pyx in zmq.backend.cython.socket.Socket.__cinit__ (zmq\backend\cython\socket.c:2365)()

TypeError: context must be specified
```

**Comment by tritemio** [comment]
Does anybody has any wild guess of where the problem it may be? I'm currently very busy and it may take a few day until I can work on this.

For the record the error comes from executing this notebook (it requires the [fretbursts](https://github.com/tritemio/FRETBursts) package) with lmfit from master:

http://nbviewer.ipython.org/github/tritemio/FRETBursts/blob/master/notebooks/FRETBursts%20-%20us-ALEX%20smFRET%20burst%20analysis.ipynb

I haven't had time to generate a self-contained stand-alone script that reproduces the issue yet.

**Comment by newville** [comment]
@tritemio  Sorry, no I cannot tell what is happening from the traceback alone.   Somewhere you're copying parameter values -- what are the parameter values?

Just in case it's not obvious, posting a small, clear script that demonstrates the problem is more or less necessary.   Just for the record, please don't expect me to read or use an ipython notebook.

**Comment by Tillsten** [comment]
I had the issue also with master, dont remember the details and didnt investigate because i have to finnish a poster. The fact that zmq plays a role is very strange, probaly some ipython interaction. @tritemio did you see the same error outside the notebook?

**Comment by tritemio** [comment]
@newville I agree on the self-contained script, it will come. It's just that I don't have time to work on it in the following days so I just dumped whatever information I have at this point.

**Comment by yoavram** [comment]
I had this problem too, made a self-contained script, the problem doesn't appear. So it's only in IPython notebook.
I've put py and ipynb files as [gist](https://gist.github.com/yoavram/3890c2d169b591172ab2).
This is with a master version (commit 67131a992fd60ceecaae83df7766ea57b5a00e5e) on Windows and Python 2.7, IPython 3.0.0, zmq 14.5.0.

**Comment by newville** [comment]
I'm not very familiar with how ipython work

**Comment by yoavram** [comment]
I agree that it is probably a problem with IPython. Or rather (since it's IPython 3.0.0) it might be in [Jupyter](https://github.com/jupyter).

**Comment by tritemio** [comment]
@yoavram thanks for taking the time to provide a simple stand-alone example!

**Comment by tritemio** [comment]
I did some testing and the bug appears only running lmfit master in ipython notebook 2.3, 2.4 and 3.0.

With lmfit 0.8.3 the example notebook provided by @yoavram runns without errors.

**Comment by newville** [comment]
@tritemio  just to be completely clear, can you verify that the error happens on the deepcopy:

```python 
   x = deepcopy(params)
````
at https://github.com/lmfit/lmfit-py/blob/master/lmfit/model.py#L743

I can imagine that something (zmq?) might be incorrectly assuming that deepcopy() need only work on objects that can be pickled.  The fact that it works with Python and fails with IPython really suggests that is where the problem lies.

I don't have a lot of time today to investigate, but perhaps one of you can try something.  Does this fail too?:
```python
from copy import deepcopy
from lmfit import Paramters
pars = Parameters()
pars.add('x', 1.0)
c = deepcopy(pars)
```

If that works, but @yoavram's script does not, I'm pretty confused.  If it fails, we can probably reduce that even further.

**Comment by tritemio** [comment]
@newville I get the same error by running your example.

**Comment by newville** [comment]
@tritemio -- Thanks, that should clarify the report to the IPython folks.   Parameters() does have a __deepcopy__ method (and this is much of the change from 0.8.3).    

Grasping at straws, and having tested nothing, but would you be willing to try changing https://github.com/lmfit/lmfit-py/blob/master/lmfit/parameter.py#L48  from

```python
 _pars._asteval.symtable[key] = deepcopy(val)
```

to 
```python
 _pars._asteval.symtable[key] = deepcopy(val, memo)
```

Alternatively, one could try commenting out the loop over asteval.symtable.items() (lines 47 and 48) -- that might break other things and would need testing, but it might provide a clue. 

Sorry I can't be of more help on this right now....

**Comment by tritemio** [comment]
@newville, FYI take also a look at @takluyver comments in ipython/ipython#8211.

**Comment by tritemio** [comment]
@newville I get the same error with your deepcopy substitution.

**Comment by tritemio** [comment]
I think this issue has been solved by merging #224.
